### PR TITLE
PARQUET-2094: Handle negative values in page headers

### PR DESCRIPTION
**Pull request by gszadovszky**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by plygrnd** [comment]
TravisCI runs contain multiple `ParquetDecodingException`s ([example](https://app.travis-ci.com/github/apache/parquet-mr/jobs/538955407#L2605)). Is this expected?

**Comment by gszadovszky** [comment]
> TravisCI runs contain multiple `ParquetDecodingException`s ([example](https://app.travis-ci.com/github/apache/parquet-mr/jobs/538955407#L2605)). Is this expected?

@plygrnd, yes this is expected. The test class [TestCorruptThriftRecords](https://github.com/apache/parquet-mr/blob/master/parquet-thrift/src/test/java/org/apache/parquet/hadoop/thrift/TestCorruptThriftRecords.java) verifies if the correct exceptions are thrown. It uses MR to execute the tests and it seems MR logs the exceptions before they are thrown to the caller.

**Comment by plygrnd** [comment]
Okay, LGTM then.

**Comment by shangxinli** [comment]
LGTM

**Review by shangxinli — APPROVED** [review]

File: parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Utility class to validate different types of Parquet metadata (e.g. footer, page headers etc.).
+ */
+public class MetadataValidator {
+
+  /**
+   * A specific IOException thrown when invalid values are found in the Parquet file metadata (including the footer,
+   * page header etc.).
+   */
+  public static class InvalidParquetMetadataException extends IOException {
```
Comment (rdblue) [inline parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java:34]: I don't think that this is an IOException. What is the value of making this a checked exception? Why not just make this a RuntimeException? Or use some existing one like IllegalStateException or ParquetDecodingException?

File: parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      super("Metadata " + metaName + " is invalid: " + value);
+    }
+  }
+
+  static PageHeader validate(PageHeader pageHeader) throws InvalidParquetMetadataException {
+    validateValue(size -> size >= 0, pageHeader.getCompressed_page_size(), "pageHeader.compressed_page_size");
+    return pageHeader;
+  }
+
+  private static <T> void validateValue(Predicate<? super T> validator, T value, String metaName)
```
Comment (rdblue) [inline parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java:45]: Why accept a predicate? Most check methods like this use a boolean. I would expect it to work like a Precondition:

```java
if (!isValid) {
  throw new ParquetDecodingException(...);
}

int size = pageHeader.getCompressed_page_size()
validateValue(size >= 0, String.format("Compressed page size must be positive, but was: %s", size));
```

File: parquet-format-structures/src/main/java/org/apache/parquet/format/Util.java [hunk 3]
```diff
@@ -130,7 +131,7 @@ public static PageHeader readPageHeader(InputStream from) throws IOException {
 
   public static PageHeader readPageHeader(InputStream from, 
       BlockCipher.Decryptor decryptor, byte[] AAD) throws IOException {
-    return read(from, new PageHeader(), decryptor, AAD);
+    return validate(read(from, new PageHeader(), decryptor, AAD));
```
Comment (rdblue) [inline parquet-format-structures/src/main/java/org/apache/parquet/format/Util.java:134]: I think it would be more clear if you called `MetadataValidator.validate` rather than just `validate`.

File: parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java [hunk 4]
```diff
@@ -77,6 +81,21 @@ public void testReadFileMetadata() throws Exception {
...
+    PageHeader ph = new PageHeader(PageType.DATA_PAGE, 100, -50);
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    Util.writePageHeader(ph, out);
+
+    try {
+      Util.readPageHeader(in(out));
+      fail("Expected exception but did not thrown");
+    } catch (InvalidParquetMetadataException e) {
+      assertTrue("Exception message does not contain the expected parts",
+          e.getMessage().contains("pageHeader.compressed_page_size"));
```
Comment (rdblue) [inline parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java:95]: Isn't there an assertion helper so you don't need to catch the exception? Something like assertThrows in the codebase?

File: parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java (hunk 2 above)
Comment (gszadovszky) [reply parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java:45]: I am not sure why I've implemented this way. I'm fine rewriting to use a simple boolean.

File: parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java (hunk 4 above)
Comment (gszadovszky) [reply parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java:95]: Yes, there is something already implemented but in another module and I cannot use it here.

File: parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Utility class to validate different types of Parquet metadata (e.g. footer, page headers etc.).
+ */
+public class MetadataValidator {
+
+  /**
+   * A specific RuntimeException thrown when invalid values are found in the Parquet file metadata (including the
+   * footer, page header etc.).
+   */
+  public static class InvalidParquetMetadataException extends RuntimeException {
```
Comment (rdblue) [inline parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java:31]: Minor: I'd prefer it if the exception weren't an inner class since that makes it harder to reference. But this isn't a blocker.

**Review by rdblue — APPROVED** [review]

File: parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java (hunk 5 above)
Comment (gszadovszky) [reply parquet-format-structures/src/main/java/org/apache/parquet/format/MetadataValidator.java:31]: It is a fair point anyway. I'll move it before merging. Thanks a lot for the review.

File: parquet-format-structures/src/main/java/org/apache/parquet/format/InvalidParquetMetadataException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package org.apache.parquet.format;
+
+/**
+ * A specific RuntimeException thrown when invalid values are found in the Parquet file metadata (including the footer,
+ * page header etc.).
+ */
+public class InvalidParquetMetadataException extends RuntimeException {
+  <T> InvalidParquetMetadataException(String message) {
```
Comment (rdblue) [inline parquet-format-structures/src/main/java/org/apache/parquet/format/InvalidParquetMetadataException.java:27]: What is the type parameter for?

File: parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java [hunk 7]
```diff
@@ -23,13 +23,16 @@
...
+import static org.junit.Assert.fail;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
+import java.io.IOException;
 
+import org.apache.parquet.format.Util.DefaultFileMetaDataConsumer;
 import org.junit.Test;
 
-import org.apache.parquet.format.Util.DefaultFileMetaDataConsumer;
```
Comment (rdblue) [inline parquet-format-structures/src/test/java/org/apache/parquet/format/TestUtil.java:32]: Nit: unnecessary change.

**Comment by rdblue** [comment]
@gszadovszky, thanks for getting this done.
